keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1249,6 +1256,181 @@ of `processing latency` used for determining the weight of the queue may
+We will implement the above (i.e. knowledge sharing between kube-apiserver),
+if the independence assumption will not work good enough.
+The above description shows that it won't result in almost any wasted work
+if the code will be well structured.
+
+##### Cost of the watch event
+
+We assumed above that the cost of processing every watch event is equal.
+However, in practice the cost associated with sending an event consists
+of two main parts:
+- the cost of going through event change logic
+- the cost of processing the even object (e.g. deserialization or sending data
+  over network)
+The first one is close to equal independently of the event, the second one
+is more proportional to the size of the object.
+However, estimating size of the is hard to predict for PATCH or DELETE requests.
+Additionally, even for POST or PUT requests where we could potentially
+estimate it based on size of the body of the requests, we may not yet have
+access to it when we need to make a decision.
+
+One way to estimate it would be to keep a running average of watch event size
+per bucket.  While it won't give us an accurate estimate, it should amortize
+well over time.
+
+Obviously some normalization will be needed here, but it's impossible to assess
+it on paper, and we are leaving this for the implementation and tuning phase
+to figure out the details.
+
+##### Dispatching the request
+
+We described how we can estimate the cost of the request associated with
+watch events triggerred by it.  But we didn't yet said how this translates
+to dispatching the request.
+
+First of all, we need to decide how to translate the estimated cost to the
+`width` of the request and its latency.  Second, we need to introduce the
+changes to our virtual world, as the fact that the request finished doesn't
+mean that sending all associated watch events has also finished (as they are
+send asynchronously).
+
+Given that individual watch events are to significant extent processed
+independently in individual goroutines, it actually makes sense to adjust the
+`width` of the request based on the expected number of triggerred events.
+However, we don't want to inflate the width of every single request that is
+triggering some watch event (as described in the above sections, setting
+the width to greater is reducing our ability to fully utilize our capacity).
+The exact function to computing the width should be figured out during
+further experiments, but the initial candidate for it would be:
+```
+  width(request) = min(floor(expected events / A), concurrency units in PL)
+```
+
+However, adjusting the width is not enough because, as mentioned above,
+processing watch events happens asynchronously.  As a result, we will use
+the mechanism of `additional latency` that is described in the section
+about LIST requests to compensate for asynchronous cost of the request
+(which in virtual world quals `<width> x <additional latency>`.
```

Comment:
s/quals/equals/

But per my comments on the other PR, I don't think the queues should be doing this multiplication. :)

Reply:
Let's try to resolve this on LIST PR (and I will adjust here later if needed).

Reply:
Sure.

Perhaps one reason to weight a bit more towards adding more latency than width is that it will slow down repeated modifications to the same object more.

Reply:
ok we resolved it over there.